L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,387 @@
+* It is possible to split any pipeline of Publisher operations into separate functions or even separate RPCs. Backpressure, cancellation and error handling still work transparently.
+* Splitting the plumbing code from logic like this can help make code more readable and testable.
+
+
+#### More
+
+More examples are possible, but this is already getting long. Please ask for more if you want to see what a particular use case would look like. Omitted examples include:
+
+* **Bidirectional streaming:** It is just client and server streaming combined. The handler takes a Publisher and returns a Publisher that can emit more than one element.
+* **Making two upstream calls in parallel:** Can be done with the [`Zip` operator](https://github.com/per-gron/shuriken/blob/master/src/rs/doc/reference.md#zippublisher).
+* **Cancellation:** Most of the time no explicit code related to cancellation is needed. For example, if an RPC has made two upstream requests in parallel and one fails, the other is automatically cancelled. This works with timeouts and streams, making two calls and using just the fastest result ([`Merge()`](https://github.com/per-gron/shuriken/blob/master/src/rs/doc/reference.md#mergepublisher)+[`Take(1)`](https://github.com/per-gron/shuriken/blob/master/src/rs/doc/reference.md#takecount)) etc as well.
+* **Error handling:** With errors, Publishers behave very much like futures/promises do in most languages/implementations. It is basically an async version of `try`/`catch`. Errors are automatically propagated through the chain of operators until they are caught, usually with the `Catch` operator.
+
+
+#### Threading model
+
+The default threading model for a server that uses this library is:
+
+* The library spawns one thread/`CompletionQueue` per CPU core.
+* When a request arrives, it is pegged to the `CompletionQueue` that it arrived at: If the request handler makes upstream requests, the responses are handled on the same `CompletionQueue`.
+* If a request handler does heavy work, it is responsible for doing that somewhere else than on the threads created by the library; in order to be able to jump back the library provides a `Schedule` function that asynchronously executes a callback on the original `CompletionQueue` thread.
+
+This way, an application will be able to utilize multiple threads, yet no synchronization primitives are needed in the rs library, and they are needed in application code only if it has to do slow work or if requests need to interact.
+
+
+## Rationale
+
+Alternate approaches for a C++ for gRPC include
+
+* **The current synchronous API:** Too slow for many use cases.
+* **The current asynchronous API:** Too low level for many use cases.
+* **Futures, as suggested in [issue #7352](https://github.com/grpc/grpc/issues/7352):** Futures typically do not have a concept of cancellation and they do not map naturally to streams.
```

Comment:
We have a more fleshed out proposal for futures internally, and cancellation is fully supported. Hoping @vjpai can take some time to publish it soon.

Reply:
That sounds very interesting! I would love to read that.